### ISSUE TITLE
Questions about image normalization

### ISSUE THREAD

**Issue by DingkunLiu** [matterport/Mask_RCNN#231]
I check image preprocessing steps in model.py and only found mean subtraction. Do we need to further divide the image by its std? 
I also checked the keras imagenet pretrained model. For tensorflow model the normalization steps are divide by 127.5 and subtract 1 for each channel. So do we need to maintain the same preprocessing steps when we use imagenet pretrained model?

In model.py 
line 2504: function mold_image
line 2512: function unmold_image

**Comment by John1231983** [comment]
Interesting also! AFAIK, the normalization must be same as how the pre-trained model did. Can anyone give us the answer?

**Comment by DingkunLiu** [comment]
Hi, John! in [deep-learning-models/resnet50](https://github.com/fchollet/deep-learning-models/blob/master/resnet50.py#L295) it uses preprocessing function from keras application [](https://github.com/fchollet/deep-learning-models/blob/master/resnet50.py#L33), which is different from the code you referenced above [](https://github.com/keras-team/keras/blob/master/keras/applications/imagenet_utils.py#L41).

**Comment by John1231983** [comment]
Sorry, You are right. I referred the above code because I download the pre-trained model from this one. Sorry about that

**Comment by DingkunLiu** [comment]
Yeah, I think models trained in this repository actually didn't use the [function](https://github.com/fchollet/deep-learning-models/blob/master/imagenet_utils.py#L23). In those models they directly import preprocessing function from keras applications. I think maybe it is an out-of-date file.

**Comment by John1231983** [comment]
Thanks. Do you know where can I download current pre-trained model that used normalization image with zero mean and unit variance?

**Comment by John1231983** [comment]
I found it. But it same as the repository did. So I guess the image must be normalization as you said. Maybe some bug we have detected

**Comment by DingkunLiu** [comment]
You remind me to check the released code to train the resnet50. It used the preprocessing function you mentioned above. So actually the implementation is right here.

**Comment by John1231983** [comment]
@DingkunLiu : Could you tell me where did you find. I still refer to official keras code and it used the normalization `preprocess_input` that has zero mean and unit variance.
```
if mode == 'tf':
        x /= 127.5
        x -= 1.
        return x
```
 While the implementation just subtraction as

`return images.astype(np.float32) - config.MEAN_PIXEL`

**Comment by John1231983** [comment]
I see, the default is `caffe`. So, it just substract mean. However, if we use caffe style, the image order also need to change to channels first. But the implementation used channels last

**Comment by DingkunLiu** [comment]
No it is not the case. The code you see hasn't been released, so you need to check the released v0.2. In that version it used function from its util and it only subtract mean from each channel.

**Comment by John1231983** [comment]
This is code in v0.2
```
def mold_image(images, config):
    """Takes RGB images with 0-255 values and subtraces
    the mean pixel and converts it to float. Expects image
    colors in RGB order.
    """
    return images.astype(np.float32) - config.MEAN_PIXEL
```

So, it only substract mean value as caffe, But I am asking about channel order. RGB or BGR

**Comment by DingkunLiu** [comment]
In the weights (download link)[ https://github.com/fchollet/deep-learning-models/archive/v0.2.zip] you can check the released code.

发送自 Windows 10 版邮件应用

发件人: John1231983
发送时间: 2018年2月5日 18:55
收件人: matterport/Mask_RCNN
抄送: Dingkun.Liu; Mention
主题: Re: [matterport/Mask_RCNN] Questions about image normalization (#231)

@DingkunLiu : Could you tell me where did you find. I still refer to official keras code and it used the normalization preprocess_input that has zero mean and unit variance.
if mode == 'tf':
        x /= 127.5
        x -= 1.
        return x
While the implementation just subtraction as
return images.astype(np.float32) - config.MEAN_PIXEL

**Comment by John1231983** [comment]
I have downloaded it and the code. I still have the problem with the image order channels first or channels last. I think it must be also convert to channel first when using caffe style
